stablehlo/dialect/StablehloOps.cpp

```diff
@@ -4860,80 +4860,62 @@ void SortOp::build(OpBuilder& builder, OperationState& state,
+    MLIRContext*, Optional<Location> location, ValueShapeRange operands,
+    DictionaryAttr attributes, RegionRange regions,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  SortOp::Adaptor adaptor(operands, attributes, regions);
+  auto operandTypes = adaptor.operands().getTypes();
+  for (auto operandType : operandTypes) {
+    auto operandShapedType = operandType.cast<ShapedType>();
+    if (operandShapedType.hasRank()) {
+      int64_t cmpDim = adaptor.dimension();
+      int64_t rank = operandShapedType.getRank();
+      if (cmpDim < -rank || cmpDim >= rank)
+        return emitOptionalError(
+            location, "dimension attribute value must be in range [-", rank,
+            ", ", rank, "), but found ", cmpDim);
+      else
+        break;  // ODS SameOperandsAndResultShape asserts inputs have same shape
+    }
+  }
 
-  for (const auto& indexedOperand : llvm::enumerate(operands)) {
-    int index = indexedOperand.index();
+  // Comparator must have 2 * N scalar arguments of same type as the N inputs.
+  Block& block = adaptor.comparator().front();
+  size_t numOperands = operandTypes.size();
+  if (block.getNumArguments() != 2 * numOperands)
+    return emitOptionalError(location, "comparator block should have ",
+                             2 * numOperands, " arguments");
+  for (const auto& indexedOperandType : llvm::enumerate(operandTypes)) {
+    int index = indexedOperandType.index();
     Type elementType =
-        indexedOperand.value().getType().cast<ShapedType>().getElementType();
+        indexedOperandType.value().cast<ShapedType>().getElementType();
     Type tensorType = RankedTensorType::get({}, elementType);
     for (int i : {2 * index, 2 * index + 1}) {
       Type argType = block.getArgument(i).getType();
       if (argType != tensorType)
-        return emitOpError("comparator block argument #")
-               << i << " should be of type " << tensorType << " but got "
-               << argType;
+        return emitOptionalError(location, "comparator block argument #", i,
+                                 " should be of type ", tensorType, " but got ",
+                                 argType);
     }
   }
 
-  // Mapped computation must return single output.
+  // Comparator must return single 0-ranked tensor with element-type i1.
   auto comparatorResult = block.getTerminator()->getOperands();
   if (comparatorResult.size() != 1)
-    return emitOpError() << "comparator must return single output, but got: "
-                         << comparatorResult.size();
-
-  // The output of computation must be 0-ranked tensor with element-type i1.
+    return emitOptionalError(location,
+                             "comparator must return single output but got ",
+                             comparatorResult.size());
   auto comparatorResultType =
       comparatorResult[0].getType().dyn_cast<RankedTensorType>();
   if (!comparatorResultType || comparatorResultType.getRank() != 0 ||
```

Comment:
Same comment as in the nearby PR for `stablehlo.while`. If the comparator returns `tensor<*xi1>`, why is that an error?

Reply:
Fixed in the similar way like https://github.com/openxla/stablehlo/pull/207